### PR TITLE
Adapt Queries & Refactor

### PR DESCRIPTION
**Pull request by sebastiancaraballo**
## What does this PR do?
- This PR aim is to start querying metrics just in time and stop processing them beforehand using jobs and saving them at `Metric` table.
- More specifically, it starts using `MergeTime` and `CompletedReviewTurnaround` tables.
- Since in the past these jobs created `Metrics`, when returning metrics for graphic representation, data was already split in weeks. However, since now these metrics are calculated just in time, the time range is now divided by week so it keeps returning the same result.

### Refactor
- Refactor `app/services/metrics/merge_time/*` and `app/services/metrics/review_turnaround/*` because now join queries are bringing unnecessary data which are then filtered at chartkick builders. E.g. `Metrics::ReviewTurnaround::PerProject` class should allow filtering by project, instead of joining all projects. 
- `app/services/metrics/merge_time/*` and `app/services/metrics/review_turnaround/*` `process` methods have the same structure, which may be refactored as well.

Resolves [#EM-209](https://rootstrap.atlassian.net/browse/EM-209?atlOrigin=eyJpIjoiNTBmNGMzZDc3ZGM3NDg1Zjg0MmM2MzhiZmFkMDBkMWMiLCJwIjoiaiJ9)

### CODE REVIEW

File: app/services/metrics/merge_time/per_department.rb [hunk 1]
```diff
@@ -4,21 +4,30 @@ class PerDepartment < Metrics::BaseDevelopmentMetrics
...
 
       def process
-        project_metrics_per_department.each do |department_id, merge_times_value|
-          create_or_update_metric(department_id, Department.name, merge_time_interval,
-                                  merge_times_value, :merge_time)
+        metrics = []
+        week_intervals = split_in_weeks(metric_interval)
+        week_intervals.map do |week|
+          interval = build_interval(week)
+          project_metrics_per_department(interval).each do |department_id, metric_value|
```
Comment (hdamico) [inline app/services/metrics/merge_time/per_department.rb:11]: You could use an `.inject({})` instead of using `metrics = []`

Comment (hdamico) [reply app/services/metrics/merge_time/per_department.rb:11]: Or `each_with_object({})`

File: app/services/builders/chartkick/language_data.rb [hunk 2]
```diff
@@ -2,16 +2,23 @@ module Builders
   module Chartkick
     class LanguageData < Builders::Chartkick::Base
       def call
-        retrieve_department_languages_metrics.map do |language|
-          metrics = language.metrics.where(@query)
-          { name: language.name, data: build_data(metrics) }
+        metrics.group_by(&:ownable_id).map do |language_metrics|
+          language = Language.find(language_metrics.first)
+          { name: language.name, data: build_data(language_metrics.second) }
```
Comment (hdamico) [inline app/services/builders/chartkick/language_data.rb:7]: Why did you use `language_metrics.second`?

File: app/services/metrics/base.rb [hunk 3]
```diff
@@ -1,28 +1,26 @@
...
-    def calculate_metrics_avg(entities, metric_type)
-      entities.reject { |_entity, count| count == 1 }.each do |entity, count|
-        Metric.find_by!(ownable: entity, value_timestamp: metric_interval, name: metric_type)
-              .tap do |metric|
-          metric.value = metric.value / count
-          metric.save!
-        end
-      end
-    end
+    Metric = Struct.new(:ownable_id, :ownable_type, :value_timestamp, :name, :value)
```
Comment (hdamico) [inline app/services/metrics/base.rb:3]: What is this for?

File: app/services/metrics/merge_time/per_language.rb [hunk 4]
```diff
@@ -4,21 +4,30 @@ class PerLanguage < Metrics::BaseDevelopmentMetrics
...
 
       def process
-        project_metrics_per_language.each do |language_id, merge_times_value|
-          create_or_update_metric(language_id, Language.name, merge_time_interval,
-                                  merge_times_value, :merge_time)
+        metrics = []
+        week_intervals = split_in_weeks(metric_interval)
+        week_intervals.map do |week|
+          interval = build_interval(week)
+          project_metrics_per_language(interval).each do |language_id, metric_value|
```
Comment (hdamico) [inline app/services/metrics/merge_time/per_language.rb:11]: Same here [comment](https://github.com/rootstrap/rs-code-review-metrics/pull/392/files#r551328144)

File: app/services/metrics/base.rb (hunk 3 above)
Comment (sebastiancaraballo) [reply app/services/metrics/base.rb]: I use structs to return results in a defined way. In the last commit, I delete some of the attributes.

File: app/services/metrics/merge_time/per_language.rb [hunk 5]
```diff
@@ -1,34 +1,25 @@
...
-          project_metrics_per_language(interval).each do |language_id, metric_value|
-            metrics << Metric.new(language_id,
-                                  Language.name,
-                                  interval.first,
-                                  :merge_time,
-                                  metric_value)
+          query(interval).each do |language_id, metric_value|
+            accumulator << Metric.new(language_id, interval.first, metric_value)
           end
+          accumulator
```
Comment (hdamico) [inline app/services/metrics/merge_time/per_language.rb]: Here you can do it even more simple:
```
        week_intervals.flat_map do |week|	
          interval = build_interval(week)
          query(interval).map do |project_id, metric_value|
            Metric.new(project_id, interval.first, metric_value)
          end
        end
```

File: app/services/metrics/merge_time/per_department.rb [hunk 6]
```diff
@@ -1,25 +1,25 @@
...
       def process
-        project_metrics_per_department.each do |department_id, merge_times_value|
-          create_or_update_metric(department_id, Department.name, merge_time_interval,
-                                  merge_times_value, :merge_time)
+        week_intervals.each_with_object([]) do |week, accumulator|
+          interval = build_interval(week)
+          query(interval).each do |department_id, metric_value|
+            accumulator << Metric.new(department_id, interval.first, metric_value)
+          end
+          accumulator
```
Comment (hdamico) [inline app/services/metrics/merge_time/per_department.rb:12]: Here you can do it even more simple:
```
        week_intervals.flat_map do |week|	
          interval = build_interval(week)
          query(interval).map do |project_id, metric_value|
            Metric.new(project_id, interval.first, metric_value)
          end
        end
```

**Review by hdamico — APPROVED** [review]

**Review by nicofh — APPROVED** [review]

**Pull request merged**